### PR TITLE
fix: Pool Coin Decimal Truncation During Deposit

### PR DESCRIPTION
**Pull request by dongsam**
## Description

closes: #445

For now, it is a draft version that reproduces the above issue and solves it at the PoC level, requiring more testing and edge case simulation.

## Tasks
- [x] Fix `ExecuteDeposit` function
- [x] Add test code to verify that the issue has been resolved well
- [x] Update `MintingPoolCoinsInvariant` and `DepositInvariant`
- [ ] Simulation to find edge cases
- [x] Remove print codes temporary written for debugging

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: x/liquidity/keeper/invariants.go [hunk 1]
```diff
@@ -94,10 +95,15 @@ func MintingPoolCoinsInvariant(poolCoinTotalSupply, mintPoolCoin, depositCoinA,
...
-	if depositCoinARatio.LT(poolCoinRatio) || depositCoinBRatio.LT(poolCoinRatio) {
-		panic("invariant check fails due to incorrect ratio of pool coins")
+	// NewPoolCoinAmount / LastPoolCoinSupply == AfterRefundedDepositCoinA / LastReserveCoinA
+	// NewPoolCoinAmount / LastPoolCoinSupply == AfterRefundedDepositCoinA / LastReserveCoinB
+	if depositCoinA.GTE(coinAmountThreshold) && depositCoinB.GTE(coinAmountThreshold) &&
+		lastReserveCoinA.GTE(coinAmountThreshold) && lastReserveCoinB.GTE(coinAmountThreshold) &&
+		mintPoolCoin.GTE(coinAmountThreshold) && poolCoinTotalSupply.GTE(coinAmountThreshold) {
+		if errorRate(depositCoinARatio, poolCoinRatio).GT(errorRateThreshold) ||
+			errorRate(depositCoinBRatio, poolCoinRatio).GT(errorRateThreshold) {
+			panic("invariant check fails due to incorrect ratio of pool coins")
```
Comment (dongsam) [inline x/liquidity/keeper/invariants.go:105]: @hallazzang Could you add a testcase for this panic for coverage?

Comment (dongsam) [reply x/liquidity/keeper/invariants.go:105]: of if there are test codes for the case already, but not detected on codecov, we can merge https://github.com/tendermint/liquidity/pull/451

File: x/liquidity/keeper/liquidity_pool.go [hunk 2]
```diff
@@ -240,77 +237,53 @@ func (k Keeper) ExecuteDeposit(ctx sdk.Context, msg types.DepositMsgState, batch
...
-		outputs = append(outputs, banktypes.NewOutput(reserveAcc, acceptedCoins))
+	poolCoinTotalSupply := k.GetPoolCoinTotalSupply(ctx, pool)
+	poolCoinMintAmt := sdk.MinDec(
+		poolCoinTotalSupply.ToDec().MulTruncate(depositCoinA.Amount.ToDec()).QuoTruncate(lastReserveCoinA.Amount.ToDec()),
+		poolCoinTotalSupply.ToDec().MulTruncate(depositCoinB.Amount.ToDec()).QuoTruncate(lastReserveCoinB.Amount.ToDec()),
+	)
+	mintRate := poolCoinMintAmt.TruncateDec().QuoTruncate(poolCoinTotalSupply.ToDec())
+	acceptedCoins := sdk.NewCoins(
+		sdk.NewCoin(depositCoins[0].Denom, lastReserveCoinA.Amount.ToDec().Mul(mintRate).TruncateInt()),
+		sdk.NewCoin(depositCoins[1].Denom, lastReserveCoinB.Amount.ToDec().Mul(mintRate).TruncateInt()),
```
Comment (dongsam) [inline x/liquidity/keeper/liquidity_pool.go:256]:
```suggestion
		sdk.NewCoin(depositCoins[0].Denom, lastReserveCoinA.Amount.ToDec().MulTruncate(mintRate).TruncateInt()),
		sdk.NewCoin(depositCoins[1].Denom, lastReserveCoinB.Amount.ToDec().MulTruncate(mintRate).TruncateInt()),
```
What do you think about using `MulTruncate` instead of `Mul` here? @hallazzang @typark391 @kogisin 
I think both will have the same result but that is more intuitive, or is there a case that is not?

File: x/liquidity/keeper/liquidity_pool_test.go [hunk 3]
```diff
@@ -173,6 +173,143 @@ func TestExecuteDeposit(t *testing.T) {
...
+	depositorPoolCoin := simapp.BankKeeper.GetBalance(ctx, depositor, pool.PoolCoinDenom)
+	require.True(sdk.IntEq(t, sdk.OneInt(), depositorPoolCoin.Amount))
+	require.True(t, simapp.BankKeeper.GetBalance(ctx, depositor, denomA).Amount.IsPositive())
+	require.True(t, simapp.BankKeeper.GetBalance(ctx, depositor, denomB).Amount.IsPositive())
+
+	// Withdraw.
+	_, err = simapp.LiquidityKeeper.WithdrawWithinBatch(ctx, types.NewMsgWithdrawWithinBatch(depositor, pool.Id, depositorPoolCoin))
+	require.NoError(t, err)
+	liquidity.BeginBlocker(ctx, simapp.LiquidityKeeper)
+	liquidity.EndBlocker(ctx, simapp.LiquidityKeeper)
```
Comment (dongsam) [inline x/liquidity/keeper/liquidity_pool_test.go:274]: Is there a reason why `BeginBlocker` was done before `EndBlocker`? The normal batch message flow is `BeginBlocker`->`*WithinBatch`(msg level)->`EndBlocker`. @hallazzang

**Comment by dongsam** [comment]
Logic `ExecuteDeposit` was simplified, so the hit line was reduced, even codecov failed, and the actual functional coverage was not lowered.

**Review by hallazzang — APPROVED** [review]

File: x/liquidity/keeper/liquidity_pool.go (hunk 2 above)
Comment (hallazzang) [reply x/liquidity/keeper/liquidity_pool.go]: Using `MulTruncate` here can cause(very rarely) a depositor to take advantage against the pool.
Since the result of `MulTruncate` is always less or equal than the result of `Mul`, I suggest to use `Mul` here.
How do others think?

Comment (dongsam) [reply x/liquidity/keeper/liquidity_pool.go]: That's right. Let's keep it `Mul` first, reproduce the corresponding edge case, analyze it, and specify it on spec through other PRs.
